components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -643,20 +644,63 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
         _parser.remove_oob("OK");
         if (_busy) {
             if (_ok_received) {
+                tr_debug("send(): _ok_received.");
                 goto RETRY;
             } else if (_parser.recv("OK")) {
+                tr_debug("send(): parser found OK");
                 goto RETRY;
             }
         }
+        // This means ESP8266 hasn't even started to receive data
         tr_debug("send(): Didn't get \">\"");
-        ret = NSAPI_ERROR_WOULD_BLOCK;
+        if (_sock_i[id].proto == NSAPI_TCP) {
+            ret = NSAPI_ERROR_WOULD_BLOCK;
+        } else if (_sock_i[id].proto == NSAPI_UDP) {
+            ret = NSAPI_ERROR_NO_MEMORY;
+        }
         goto END;
     }
     _ok_received = false;
     _parser.remove_oob("OK");
 
-    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
-        ret = NSAPI_ERROR_OK;
+    if (_parser.write((char *)data, (int)amount) < 0) {
+        tr_debug("Failed to write serial data");
+        // Serial is not working, serious error, reset needed.
+        ret = NSAPI_ERROR_DEVICE_ERROR;
+        goto END;
+    }
+
+    // The "Recv X bytes" is not documented.
+    int bytes_confirmed;
+    if (!_parser.recv("Recv %d bytes", &bytes_confirmed)) {
+        tr_debug("Bytes not confirmed.");
+        ret = NSAPI_ERROR_DEVICE_ERROR;
+        if (_sock_i[id].proto == NSAPI_TCP) {
+            ret = NSAPI_ERROR_WOULD_BLOCK;
+        } else if (_sock_i[id].proto == NSAPI_UDP) {
+            ret = NSAPI_ERROR_NO_MEMORY;
+        }
+        goto END;
+    } else if (bytes_confirmed != amount) {
```

Comment:
I assume `bytes_confirmed` is used as unitialized here. I would say that you should initialize to zero.